### PR TITLE
RL Minikube setup

### LINKED ISSUE #14: RL Minikube environment
**User Story**

As a(n) engineer, I will be creating a Local dev environment that aims to pair as close as possible to production-like environments. The hopes of this will enable better preparation for if and when we get to end-user adoption. The added benefits to this approach is to add more advanced experimentation with system scalability and networking.

**Acceptance Criteria**

An up and running/functioning Record-Linkage app and MPI, within a kubernetes cluster, running in Minikube

### PR DESCRIPTION
**Pull request by derekadombek**
# Pull Request

## Description
A local dev environment that aims to pair as close as possible to production-like environments. The hopes of this will enable better preparation for if and when we get to end-user adoption. The added benefits to this approach is to add more advanced experimentation with system scalability and networking.

**NOTE**: If you try to run this environment from this repo, it probably wont work because this is intended for testing the source code in `phdi`

## Related Issues
closes #14 

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [ ] I have minimized the number of reviewers to include only those essential for the review.
- [x] I have notified teammates in the review thread to build awareness.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: dedupe/manifests/scripts/minikube-start.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+cd $(dirname $0)
+minikube delete
+minikube start
+eval $(minikube docker-env)
+
+if [ -z $1 ];
+then
+    docker build -t phdi/record-linkage ../../
```
Comment (ericbuckley) [inline dedupe/manifests/scripts/minikube-start.sh:8]: @derekadombek does this step assume you have the [phdi repo](https://github.com/CDCgov/phdi) checked out locally and in the same directory as the IDWA repo?

File: dedupe/manifests/record-linkage.yml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  PGDATA: /data/pgdata
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: postgres-db-lb
+spec:
+  selector:
+    app: postgresql-db
+  type: LoadBalancer
```
Comment (ericbuckley) [inline dedupe/manifests/record-linkage.yml:21]: I'm a bit new to kubernetes, but is the database loadbalancer just needed to forward the postgres service to the host?

File: dedupe/manifests/record-linkage.yml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    matchLabels:
+      app: record-linkage
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: record-linkage
+    spec:
+      initContainers:
+        - name: await-db-ready
```
Comment (ericbuckley) [inline dedupe/manifests/record-linkage.yml:113]: I see a postgres service created on line 69, whats the difference between that one and this

**Review by ericbuckley — COMMENTED** [review]
Thanks for doing this, its giving me a crash course in kubernetes!

File: dedupe/manifests/record-linkage.yml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+kind: ConfigMap
+metadata:
+  name: postgres-db-config
+  labels:
+    app: postgresql-db
+data:
+  POSTGRES_DB: testdb
+  POSTGRES_USER: postgres
+  POSTGRES_PASSWORD: pw
+  PGUSER: postgres
```
[the remark this answers is not in the document]
Comment (derekadombek) [reply dedupe/manifests/record-linkage.yml:11]: That's a great question, I believe it was included with the docker-compose file so I just threw it in. I'm gonna see if it's actually being used.

File: dedupe/manifests/record-linkage.yml (hunk 2 above)
Comment (derekadombek) [reply dedupe/manifests/record-linkage.yml:21]: Correct! So in k8s there are 3 options to choose from, ClusterIP - which is only internal to the cluster and allows automatic communication between resources(also, ClusterIP is enabled during the other two types as well, for good reason), and then theres NodePort and Loadbalancer. They both essentially do the same things which is connecting with the Node to create an external IP. The main difference is that Loadbalancer is the best option for cloud provider connectivity, to then do other things from there like dns, certificates, Ingress and egress, etc. tbh we probably should just to NodePort here because this is just a local sandbox environment anyways but I'm trying to keep it as close to the real thing as possible. 

So once any of those are set. There are certain ways to communicate with them from CLI, without getting too into that, to summarize, if we were to use ClusterIP we could need to do whats called port-fowarding for each service, every time we want to them. For NodePort and Loadbalancer, it depends on the provider/platform your are using but in our case we are using Minikube, in the script we use the command `minikube tunnel` and that will expose an external IP for you to use when running `kubectl get all`. If you dont use the tunnel the external-IP will show up as `<Pending>`.

Sorry for the brain dump on Kube lol

File: dedupe/manifests/record-linkage.yml (hunk 3 above)
Comment (derekadombek) [reply dedupe/manifests/record-linkage.yml:113]: This step is definitely a unique one in regards to our "Stateful" type design. Kubernetes will just go and provision everything basically all at the same time. Thats okay and quick for some things but not necessarily for databases that need to do migrations which naturally just take time. Thats where `InitContainers` come in handy. Basically if we didn't use initContainers here the api deployment we fail fast and in a broken state due to health check on when communicating with the db service. So, what this does is pauses the api service from doing anything until the db is done getting ready for it.

File: dedupe/manifests/scripts/minikube-start.sh (hunk 1 above)
Comment (ericbuckley) [reply dedupe/manifests/scripts/minikube-start.sh:8]: The else block did work for me, are you saying that the else block won't work for individuals running docker on apple silicon?

Comment (derekadombek) [reply dedupe/manifests/scripts/minikube-start.sh:8]: ohhh! Thats great to hear! I was unable to get it to work even after figuring out the Rosetta thing yesterday. I even ran different arch minikubes. Did you install the arm64 or the amd64 minikube?

File: dedupe/manifests/record-linkage.yml (hunk 3 above)
Comment (ericbuckley) [reply dedupe/manifests/record-linkage.yml:113]: Another option you could consider...The API container has [pyway](https://github.com/jasondcamp/pyway) installed to run the database migrations, so you could likely use the `pyway validate` command to check to see if the connection is ready in a container readinessProbe.  Outcome should be the same, but you'd have 1 less container.

```sh
pyway --database-type postgres --database-host ${POSTGRES_DB_LB_SERVICE_HOST} --database-port 5432 --database-name ${POSTGRES_DB} --database-username ${POSTGRES_USER} --database-password ${POSTGRE_PASSWORD} --database-migration-dir migrations/ validate
```

File: dedupe/manifests/scripts/minikube-start.sh (hunk 1 above)
Comment (ericbuckley) [reply dedupe/manifests/scripts/minikube-start.sh:8]: amd64, I'm using a 2019 intel macbook pro

Comment (derekadombek) [reply dedupe/manifests/scripts/minikube-start.sh:8]: ah that seems to make sense then. your docker manifest arch inside your minikube instance probably matches up properly to the image arch. I noticed when trying this yesterday that my manifest wouldnt change to match it, even while using the amd64 install. also minikube really loves to complain when I try to use that install haha

File: dedupe/manifests/record-linkage.yml (hunk 3 above)
Comment (derekadombek) [reply dedupe/manifests/record-linkage.yml:113]: Oh yeah thats not a bad idea!! i like it. From my understanding with Initcontainers, they are no longer around after they are complete BUT that idea maybe be more efficient in regards to startup. Something to try for sure.

Comment (derekadombek) [reply dedupe/manifests/record-linkage.yml:113]: I actually got this is work with just the health check endpoint `/` on the `readinessProbe`. Its something I actually tried doing early on into this setup but I must have missed on using delay times or something and never got it to work then. Happy it does now though. I'm about to push it up. If you could please double check next week or something on your end it works that would be awesome :)
